src/deepsparse/transformers/utils/decoder_kv_cache.py

```diff
@@ -0,0 +1,218 @@
+
+
+class DecoderKVCache:
+    def __init__(self, engine_type: str):
+        """
+        The goal this object is to handle the manipulation
+        of the key value cache.
+
+        :param engine_type: The engine type to use for the decoder.
+            If the engine type is DEEPSPARSE_ENGINE, then the
+            KVCache object from the deepsparselib will be loaded as
+            an attribute. This object is used to handle the manipulation
+            of the key value buffers on the DeepSparse engine side.
+        """
+
+        # assuming that kv cache arrays are of shape
+        # [batch_size, num_heads, sequence_length, hidden_size]
+        self._sequence_axis = 2
+        self._session_id = None
+        self._freeze_first_position = None
+        self._state = None
+        self._kv_cache = KVCache() if engine_type == DEEPSPARSE_ENGINE else None
+
+    def setup_session(
+        self,
+        session_id: str,
+        state: Dict[str, Any],
+        num_tokens: int,
+        freeze_first_position: bool = False,
+    ):
+        """
+        Setup the session - a level of abstraction that allocates
+        the resources to store and manipulate the kv cache.
+
+        :param session_id: The session id to use for the current
+            session. Used to identify the kv cache state
+        :param state: The state of the cache. This is a dictionary
+            that maps the name of the cache array to the cache array.
+            The cache tensor is a numpy array of shape
+            [batch_size, num_heads, sequence_length, hidden_size]
+        :param num_tokens: The number of tokens processed so far,
+            corresponding to the number of "non-blank" entries in the
+            kv cache array. Even though all numpy arrays in the state
+            of the cache have constant sequence length, this length
+            corresponds to the total "capacity" of the
+            cache. A portion of this memory can be "blank".
+        :param freeze_first_position: If set to True, once the kv cache
+            gets filled, the position along the sequence length axis
+            that corresponds to the first token will be frozen.
+            This assures that, once the KV cache is full (there are no
+            "blank" entries), and we are removing the "oldest" entry
+            from the cache, we will nevertheless keep the cache entry
+            that corresponds to the BOS token in the sequence.
+            By default, is set to False.
+        """
+        self.session_id = session_id
+        self._state = state
+        # TODO: I am aware that I am altering the specs with this
+        # variable, but it makes more sense to set this as a boolean
+        # rather than a list of integers.
+        self._freeze_first_position = freeze_first_position
+        if self._kv_cache is not None:
+            self._kv_cache.reset(
+                num_tokens=num_tokens, freeze_first_position=self._freeze_first_position
```

Comment:
here num_tokens is NOT the number of actual tokens processed so far, this is the entire kv cache "capacity". Fetch this from the state.